sql/core/src/test/scala/org/apache/spark/sql/sources/BucketedReadSuite.scala

```diff
@@ -735,4 +744,88 @@ abstract class BucketedReadSuite extends QueryTest with SQLTestUtils {
+  test("SPARK-27100 stack overflow: read non bucketed data with large partitions") {
+    testLargeFilePartitionStackOverflow(true)
+  }
+
+  private def testLargeFilePartitionStackOverflow ( isBucketed : Boolean) {
+    // Need a large number of files in the partition for the overflow
+    val numFilesInPartition = 100000
+    val partitionValues = InternalRow.apply(Array("a"))
+
+    val schema = new StructType()
+    val fakeHadoopFsRelation = new HadoopFsRelation(null, schema, schema, null, null, null)(spark)
+    val optionalBucketSet = null
+    val bucketSpec = new BucketSpec(1, Seq("a"), Seq("b"))
+    val files = (0 to numFilesInPartition).toStream.map { i =>
+      new FileStatus(10, false, 1, 512, 1000,
+        new Path(s"file${i}_0.zzz"))
+    }
+    val partitionDirectory = PartitionDirectory(partitionValues, files);
+
+    val fileSource =
+     FileSourceScanExec(fakeHadoopFsRelation,
+      null,
+      schema,
+      null,
+      Option(optionalBucketSet),
+      Seq.empty,
+      Option(new TableIdentifier("stackOverflow")))
+
+    val inputRDD = if (isBucketed) {
+      // Create a Bucketed RDD. This is a private method so we need to call this indirectly.
+      val createBucketedReadRDD = PrivateMethod[RDD[InternalRow]]('createBucketedReadRDD)
+
+      fileSource invokePrivate createBucketedReadRDD(bucketSpec,
+        (file: PartitionedFile) => Seq(InternalRow(1)).toIterator,
+        Array(partitionDirectory),
+        fakeHadoopFsRelation)
+    } else {
+      // Create a Bucketed RDD. This is a private method so we need to call this indirectly.
+      val createNonBucketedReadRDD = PrivateMethod[RDD[InternalRow]]('createNonBucketedReadRDD)
+
+      fileSource invokePrivate createNonBucketedReadRDD(
+        (file: PartitionedFile) => Seq(InternalRow(1)).toIterator,
+        Array(partitionDirectory),
+        fakeHadoopFsRelation)
+
+    }
+    // check to make sure we've created the a big enough file partition.
+    // also guarantees that the 'files' Stream is initialized before we
+    // attempt to serialize it in the task.
+    val count = inputRDD.partitions(0).asInstanceOf[FilePartition].files.length;
+    assert(count == numFilesInPartition + 1)
+
+    // Create a task encapsulating the FilePartition
+    val task = new ShuffleMapTask(0, 0,
+      null, inputRDD.partitions(0), Seq(TaskLocation("host0", "execA")), new Properties, null)
+    // Serialize the task and catch the exception
+    val env = SparkEnv.get
+    val ser = env.closureSerializer.newInstance()
+    try {
+      ser.serialize(task)
+    } catch {
+      case ex: StackOverflowError =>
+        val bucketingType = if (isBucketed) {
+          "bucketed"
+        } else {
+          "non-bucketed"
+        }
+        fail("Stack Overflow Exception in serializing task to read partitioned %s tables"
+          .format(bucketingType))
+      case _ => fail("Exception in serializing task to read partitioned tables")
```

Comment:
This would mask the exception. Just let the exception go.

Reply:
Done